### DISCUSSION TITLE
Background subtraction

### DISCUSSION THREAD

**Discussion by 1nelly1** [ImageMagick/ImageMagick#2295] [Help]
Hello,
I would like to receive a mask for an image showing an object. In order to do this automatically I took a second image showing only the background. I have version 7.0.10-23 Q16 on Windows 10.
I tried:
convert DGH_2990.JPG DGH_2989.JPG -compose difference -composite -separate -evaluate-sequence add difference.jpg
But there are still a lot of unwanted pixels showing.
I tried -connected-components in order to remove them but I get the message "magick: too many objects".
Here are the links to both pictures:
http://www.hollaender-online.com/download/temp/DGH_2989.JPG
http://www.hollaender-online.com/download/temp/DGH_2990.JPG
Or as raw file in NEF-file format:
http://www.hollaender-online.com/download/temp/DGH_2989.NEF
http://www.hollaender-online.com/download/temp/DGH_2990.NEF
I would appreciate if anybody can give me some ideas what steps are necessary.
Thank you
1nelly1

**Comment by snibgo** [comment]
snibgo's rule number one of image processing: Never use JPG images if better images are available. You have NEF files, so I suggest you use those.

Suggested steps:

1. Demosaic them with no white balancing or auto-brighten or anything.

```
%DCRAW% -v -W -6 -r 1 1 1 1 -g 1 0 -T -O x1.tiff DGH_2989.NEF
%DCRAW% -v -W -6 -r 1 1 1 1 -g 1 0 -T -O x2.tiff DGH_2990.NEF
```

2. If the images had different exposures, align them by darkening one. Your photos don't need this.

3. Your camera has moved slightly between exposures, so physically align one image against the other. Probably a "-distort perspective" is sufficient.

4. Now find the differences between pixels. There will be slight changes in lighting or whatever, so you will need a threshold: ignore very small differences.

5. The result is a mask, say black where there is no difference (background) and white where pixels are different (foreground object). However, some pixels in the foreground object may happen to be the same as corresponding background pixels, so this may need some cleaning.

**Reply by 1nelly1** [reply]
Dear snibgo,

thank you very much for your ideas.

I executed your two commands under 1. and experimented a little bit with "difference" and "threshold". It seems that it is even more difficult to extract the foreground by using the resulting files x1.tiff and x2.tiff.

Honestly I am a little bit frustrated. Up to now I sucessfully used  ImageMagick with Perl for simple tasks like resizing images. However I do not have deeper knowledge in image processing. I did not expect that it is a difficult task to extract the foreground if two pictures are provided. Ok, I understand that the conditions of both pictures are not 100% identical. They will never be 100% identical but I can change the conditions so that the foreground extraction is easier to execute. I was assuming that having two pictures is better for extraction of the foreground than just a green or blue screen. But maybe I am wrong. It would be strange for me but maybe I get better result by just using a green screen. Or maybe I should use both (two picures as well as green screen). What is your opinion? The two pictures I took are just for testing. What I want to do is taking something like 70-100 pictures of an object and convert that to a 3d model (photogrammetry). In oder to receive good results a mask for each picture is needed. 

I am also not sure regarding another aspect. That is if ImageMagick is really best suited for that task. Maybe OpenCV is better suited in this case. Maybe you are willing to give me your valued opinion also regarding this matter.

Thank you
1nelly1

**Comment by snibgo** [comment]
We can make a new version of x1 that spatially aligns with x2 with:
```
set sspPerspCoords=1477,322,1477,317,3774,319,3775,314,189,7187,190,7182,4479,7274,4480,7269

%IMG7%magick ^
  x1.tiff ^
  -distort Perspective %sspPerspCoords% ^
  x1p.tiff
```

Your two photos both have clipped highlights on the floor, in the green and blue channels. The lighting has changed slightly between photos, so the background also changes, especially towards the right. Clipping causes a change in hue as well as lightness.

The shadow/reflection of the object towards the camera causes a large difference. You might call this either "foreground" or "background". If it is "background", what is constant between the photos that makes it "background"?

To avoid problems with changing lighting, don't use daylight.

To avoid problems with changing movement between exposures, use multiple cameras and fire them simultaneously. For an example with 201 cameras and a living subject, see https://www.theguardian.com/artanddesign/2020/jul/15/marc-quinn-statue-colston-jen-reid-black-lives-matter-bristol

To separate foreground from background, we need a visual difference between them, and that difference should be greater than the variation within the background or foreground. A green or blue screen has a high saturation, and a fairly constant and known hue. This makes it fairly easy to detect.

> ... something like 70-100 pictures of an object and convert that to a 3d model (photogrammetry).

This is a complex requirement. IM does not contain high-level operations for that. Packages are available, for example, https://alicevision.org/#meshroom

**Comment by 1nelly1** [comment]
Dear snibgo,

thank you very much for your comments.

After your suggested alignment the result is indeed much better.
However I think that I was expecting too much and that I have to pay more attention to maintain the conditions in each photo I take. I took pictures of the same object but this time placed on a red sofa (see download links). The result I get with IM is now right away much better. 
http://www.hollaender-online.com/download/temp/DGH_3010.NEF
http://www.hollaender-online.com/download/temp/DGH_3011.NEF
Yes, to further improve I have to create a constant lighting situation and also make sure that the shadow of the object itself is not causing me headaches. Yes, I will not use daylight in the future.
Main reason for multiple cameras are moving objects like persons. I am interested in static objects so most probably I will be successfull without using many cameras.

I will use IM only for helping in creating the masks. There is - as you mentioned - other software available for finally creating the 3d model.

Thank you
1nelly1

**Comment by snibgo** [comment]
The red sofa photos are better.

```
%DCRAW% -v -W -6 -r 1 1 1 1 -g 1 0 -T -O x1.tiff DGH_3010.NEF
%DCRAW% -v -W -6 -r 1 1 1 1 -g 1 0 -T -O x2.tiff DGH_3011.NEF
```

The maximum value is about 48%, so we multiply by 2, and convert from linear RGB to non-linear sRGB. This is just to help us see the images.
```
%IMG7%magick ^
  x1.tiff ^
  -evaluate Multiply 2 ^
  -set colorspace RGB ^
  -colorspace sRGB ^
  x1s.tiff

%IMG7%magick ^
  x2.tiff ^
  -evaluate Multiply 2 ^
  -set colorspace RGB ^
  -colorspace sRGB ^
  x2s.tiff
```
We can use two methods: (a) two-photo difference and (b) green-screen subtraction.

(a) Two-photo difference

Align the photos spatially. There is no detail top-left, so I have guessed that alignment.
```
set sspPerspCoords=^
200,200,201,199,^
4570,235,4571,234,^
511,7132,513,7127,^
4420,7159,4422,7155

%IMG7%magick ^
  x1s.tiff ^
  -distort Perspective %sspPerspCoords% ^
  x1sp.tiff
```

```
%IMG7%magick ^
  x1sp.tiff x2s.tiff ^
  -compose Difference -composite ^
  -grayscale RMS ^
  -black-threshold 1%% ^
  -white-threshold 3%% ^
  d.tiff
```
d.tiff has black for background, white for foreground, and some other value (between 1% and 3%) where I'm not sure. There is noise (white spots on black background, or vice versa) that could be removed with "-despeckle" or "-morphology".


(b) Green-screen subtraction.

```
%IMG7%magick ^
  x1s.tiff ^
  -colorspace HCL ^
  -channel 2 -evaluate Set 50%% +channel ^
  -colorspace sRGB ^
  -fill Black -fuzz 4%% -opaque sRGB(64%%,44%%,41%%) ^
  x1sbf.tiff
```
We set the image to a constant lightness, 50%. The red sofa is then about sRGB(64%,44%,41%), so we make that black. We could then turn non-black into white with "-fill White +opaque Black".

This is good except for the crease in the sofa. That is about the same colour as the "chin" of the object.

We could make another mask of the sofa crease. In x1s.tiff this is about sRGB (13%,6%,4%), which is darker than the "chin". The two masks could then be combined.

Hint: cloth drapes can be arranged so they don't have creases.


Incidentally, https://www.youtube.com/user/snibgo has short movie examples of two-photo background difference with ImageMagick.

**Comment by 1nelly1** [comment]
Dear snibgo,

thank you again for your explanations and your code which I checked with the 2 sofa pictures.
As you already mentioned colours of the object can also be found on the sofa. So to create a mask automatically is difficult. That means for me that in the future I will use a background that is not having overlapping colours with the object.
Regarding these two sample pictures I found out that the following 2 lines are already producing a result that is not too bad:
convert DGH_3010.NEF DGH_3011.NEF -compose difference -composite -threshold 7.5% difference41.tiff
convert difference41.tiff -define connected-components:area-threshold=500 -define connected-components:mean-color=true -connected-components 4 difference411.tiff
I think that I could solve all my problems if the connected-components funktion would work in more complex situations. For this example it is working but if you reduce the threshold to 5% there are too many isolated pixels and the function returns an error.
There is another problem I noticed. When using a threshold  (no matter what value and also if using your described procedure) the created mask is hiding parts of the object. This refers to the left side of the object. For example the ear is much smaller. As an example I attached some details of the left side of the face.

Best regards
1nelly1

![original](https://user-images.githubusercontent.com/68476797/87993985-2c25f880-caec-11ea-85ae-b7513f578e5f.JPG)
![original_masked](https://user-images.githubusercontent.com/68476797/87993987-2d572580-caec-11ea-87ff-6145c6f139d0.JPG)
